quinn-proto/src/config.rs

```diff
@@ -308,25 +312,27 @@ where
 {
     /// Create a default config with a particular `reset_key`
     pub fn new(reset_key: S::HmacKey) -> Self {
-        Self {
-            local_cid_len: 8,
-            reset_key: Arc::new(reset_key),
-            max_udp_payload_size: MAX_UDP_PAYLOAD_SIZE,
-        }
+        let cid_factory: fn() -> Box<dyn ConnectionIdGenerator> =
+            || Box::new(RandomConnectionIdGenerator::default());
+        EndpointConfig::new_with_cid_generator(reset_key, cid_factory)
     }
 
-    /// Length of connection IDs for the endpoint.
+    /// new_with_cid_generator is designed for backward compatibility
     ///
-    /// This must be no greater than 20. If zero, incoming packets are mapped to connections only by
-    /// their source address. Otherwise, the connection ID field is used alone, allowing for source
-    /// address to change and for multiple connections from a single address. When local_cid_len >
-    /// 0, at most 3/4 * 2^(local_cid_len * 8) simultaneous connections can be supported.
-    pub fn local_cid_len(&mut self, value: usize) -> Result<&mut Self, ConfigError> {
-        if value > MAX_CID_SIZE {
-            return Err(ConfigError::OutOfBounds);
+    /// EndpointConfig can still call fn new() to use a random cid generator
+    /// new_with_cid_generator() can accept any customized cid generator that
+    /// implements ConnectionIdGenerator trait
+    pub fn new_with_cid_generator<
+        F: Fn() -> Box<dyn ConnectionIdGenerator> + Send + Sync + 'static,
+    >(
```

Comment:
I'd rather this be a setter than an additional constructor, as setters are more scalable/composable.

Reply:
The downside is that one would need a default initialization of the factory before (e.g. with the random generator factory), which would then get replaced. Introducing an `Option` and then having to deal later on with the field not being set seems not very convenient.

A `Builder` for the config could avoid this, but that's a lot bigger of an API change. 
Since the cost for the wasted `Box<RandomIdConnectionGeneratorFactory>` is likely negligible compared to the remaing setup (e.g. loading TLS keys) it's likely find to just have the setter.

Reply:
In the (indeed unlikely) event that the cost became an issue, we could even use a ZST for the default impl, which wouldn't so much as malloc.

Reply:
I agree. This only occurs at the configuration phase and thus the waste is not a big problem - Will proceed the change to setter.